zcash_client_backend/src/keys.rs

```diff
@@ -28,18 +45,223 @@ pub fn spending_key(seed: &[u8], coin_type: u32, account: u32) -> ExtendedSpendi
+) -> TransparentAddress {
+    let secp = Secp256k1::new();
+    let pk = PublicKey::from_secret_key(&secp, secret_key);
+    derive_transparent_address_from_public_key(&pk)
+}
+
+#[cfg(feature = "transparent-inputs")]
+pub fn derive_transparent_address_from_public_key(
+    public_key: &secp256k1::key::PublicKey,
+) -> TransparentAddress {
+    let mut hash160 = ripemd160::Ripemd160::new();
+    hash160.update(Sha256::digest(&public_key.serialize()));
+    TransparentAddress::PublicKey(*hash160.finalize().as_ref())
+}
+
+#[cfg(feature = "transparent-inputs")]
+pub fn derive_secret_key_from_seed<P: consensus::Parameters>(
+    params: &P,
+    seed: &[u8],
+    account: AccountId,
+    index: u32,
+) -> Result<SecretKey, hdwallet::error::Error> {
+    let private_key =
+        derive_extended_private_key_from_seed(params, seed, account, index)?.private_key;
+    Ok(private_key)
+}
+
+#[cfg(feature = "transparent-inputs")]
+pub fn derive_public_key_from_seed<P: consensus::Parameters>(
+    params: &P,
+    seed: &[u8],
+    account: AccountId,
+    index: u32,
+) -> Result<PublicKey, hdwallet::error::Error> {
+    let private_key = derive_extended_private_key_from_seed(params, seed, account, index)?;
+    let pub_key = ExtendedPubKey::from_private_key(&private_key);
+    Ok(pub_key.public_key)
+}
+
+#[cfg(feature = "transparent-inputs")]
+pub fn derive_extended_private_key_from_seed<P: consensus::Parameters>(
+    params: &P,
+    seed: &[u8],
+    account: AccountId,
+    index: u32,
+) -> Result<ExtendedPrivKey, hdwallet::error::Error> {
+    let pk = ExtendedPrivKey::with_seed(&seed)?;
+    let private_key = pk
+        .derive_private_key(KeyIndex::hardened_from_normalize_index(44)?)?
+        .derive_private_key(KeyIndex::hardened_from_normalize_index(params.coin_type())?)?
+        .derive_private_key(KeyIndex::hardened_from_normalize_index(account.0)?)?
+        .derive_private_key(KeyIndex::Normal(0))?
+        .derive_private_key(KeyIndex::Normal(index))?;
+    Ok(private_key)
+}
+
+#[cfg(feature = "transparent-inputs")]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Wif(pub String);
+
+#[cfg(feature = "transparent-inputs")]
+impl Wif {
+    pub fn from_secret_key(sk: &SecretKey, compressed: bool) -> Self {
+        let secret_key = sk.as_ref();
+        let mut wif = [0u8; 34];
+        wif[0] = 0x80;
+        wif[1..33].copy_from_slice(secret_key);
+        if compressed {
+            wif[33] = 0x01;
+            Wif(bs58::encode(&wif[..]).with_check().into_string())
+        } else {
+            Wif(bs58::encode(&wif[..]).with_check().into_string())
+        }
+    }
+}
+
+#[cfg(feature = "transparent-inputs")]
+impl<'a> TryInto<SecretKey> for &'a Wif {
+    type Error = Bs58Error;
+
+    fn try_into(self) -> Result<SecretKey, Self::Error> {
+        bs58::decode(&self.0)
+            .with_check(None)
+            .into_vec()
+            .map(|decoded| SecretKey::from_slice(&decoded[1..33]).expect("wrong size key"))
```

Comment:
This does not check that the first byte of the decoded data is 0x80 nor that the last byte is 0x01, nor that the length is 34 bytes.

(According to https://en.bitcoin.it/wiki/Wallet_import_format#WIF_to_private_key , some implementations for Bitcoin used 0x81 to 0x87 inclusive as the first bytes, but there should be no such implementations for Zcash in the wild.)

Reply:
Filed #481.

Reply:
Fixed in https://github.com/zcash/librustzcash/pull/341/commits/2e941931e67aa8a0128113d9f73ed32a7d6c0698

Reply:
@nuttycom has altered this PR to do the correct thing, but we still need to confirm that this is compatible with the mobile SDKs. I think the issues here are limited because it only affects testnet users, but we do need to document this problem and what its symptoms will look like.